### DISCUSSION TITLE
Do I need to deal with CSRF by myself?

### DISCUSSION THREAD

**Discussion by weathon** [fastapi-users/fastapi-users#622] [General]
If I am using cookie or jwt

**Comment by frankie567** [comment]
CSRF is only useful if you use Cookie authentication in a browser, since the browser automatically forwards the cookies it has in memory when making a query to your browser; potentially allowing other sites to make authenticated requests to your app maliciously.

To implement such protection, you can have a look at [`asgi-csrf`](https://github.com/simonw/asgi-csrf). It uses the double-submit cookie pattern. Basically, it'll set a CSRF token in the cookies. Then, when you make a request to your API in your frontend app, you have to set this same CSRF token in an HTTP header `x-csrftoken`. The middleware then verifies that the token in cookies is the same as the one in headers. It's safe because an external website can't directly the cookies it doesn't own to add the `x-csrftoken` header. Ref: https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#double-submit-cookie

Other things to look for if you use Cookie authentication:

* `httponly`, `secure` and `samesite` parameters for the session cookie (the defaults set by FastAPI Users should be safe): https://frankie567.github.io/fastapi-users/configuration/authentication/cookie/ It'
* CORS headers: https://fastapi.tiangolo.com/tutorial/cors/?h=cors#cors-cross-origin-resource-sharing

**Reply by weathon** [reply]
Great Thank you
